kimchi/src/circuits/polynomials/turshi.rs

```diff
@@ -0,0 +1,1025 @@
+        F::zero(),
+        F::zero(),
+        F::zero(),
+        F::zero(),
+    ]
+}
+
+impl<F: FftField> CircuitGate<F> {
+    /// This function creates a CairoClaim gate
+    pub fn create_cairo_claim(wires: GateWires) -> Self {
+        CircuitGate {
+            typ: GateType::CairoClaim,
+            wires,
+            coeffs: vec![],
+        }
+    }
+    /// This function creates a CairoInstruction gate
+    pub fn create_cairo_instruction(wires: GateWires) -> Self {
+        CircuitGate {
+            typ: GateType::CairoInstruction,
+            wires,
+            coeffs: vec![],
+        }
+    }
+
+    /// This function creates a CairoFlags gate
+    pub fn create_cairo_flags(wires: GateWires) -> Self {
+        CircuitGate {
+            typ: GateType::CairoFlags,
+            wires,
+            coeffs: vec![],
+        }
+    }
+
+    /// This function creates a CairoTransition gate
+    pub fn create_cairo_transition(wires: GateWires) -> Self {
+        CircuitGate {
+            typ: GateType::CairoTransition,
+            wires,
+            coeffs: vec![],
+        }
+    }
+
+    /// Gadget generator of the whole cairo circuits from an absolute row and number of instructions
+    pub fn create_cairo_gadget(
+        // the absolute row in the circuit
+        row: usize,
+        // number of instructions
+        num: usize,
+    ) -> Vec<Self> {
+        // 0: 1 row for final check CairoClaim gate
+        // 4i+1: 1 row per instruction for CairoInstruction gate
+        // 4i+2: 1 row per instruction for Flags argument
+        // 4i+3: 1 row per instruction for CairoTransition gate
+        // 4i+4: 1 row per instruction for Cairo Auxiliary gate
+        // ...
+        // 4n-3: 1 row for last instruction
+        // 4n-2: 1 row for Auxiliary argument (no constraints)
+        let mut gates: Vec<CircuitGate<F>> = Vec::new();
+        if num > 0 {
+            let claim_gate = Wire::new(row);
+            gates.push(CircuitGate::create_cairo_claim(claim_gate));
+        }
+        let last = num - 1;
+        for i in 0..last {
+            let ins_gate = Wire::new(row + 4 * i + 1);
+            let flg_gate = Wire::new(row + 4 * i + 2);
+            let tra_gate = Wire::new(row + 4 * i + 3);
+            let aux_gate = Wire::new(row + 4 * i + 4);
+            gates.push(CircuitGate::create_cairo_instruction(ins_gate));
+            gates.push(CircuitGate::create_cairo_flags(flg_gate));
+            gates.push(CircuitGate::create_cairo_transition(tra_gate));
+            gates.push(CircuitGate::zero(aux_gate));
+        }
+        // the final instruction
+        let ins_gate = Wire::new(row + 4 * last + 1);
+        let aux_gate = Wire::new(row + 4 * last + 2);
+        gates.push(CircuitGate::create_cairo_instruction(ins_gate));
+        gates.push(CircuitGate::zero(aux_gate));
+
+        gates
+    }
+
+    /// verifies that the Cairo gate constraints are solved by the witness depending on its type
```

Comment:
Oh that's fine, but only for `ensure_`,  since `verify_` is the one called from `gates.rs` for all other gates. Nonetheless, I can also create a `mod witness` for all things related to creating the witness, again to put some order inside `turshi.rs`, which is too long otherwise.